File: server/cluster/cluster.go
package cluster

import (
	"connectrpc.com/connect"
	"context"
	"fmt"
	"github.com/gin-gonic/gin"
	cephv1 "github.com/koor-tech/data-control-center/gen/go/api/resources/ceph/v1"
	clusterpb "github.com/koor-tech/data-control-center/gen/go/api/services/cluster/v1"
	"github.com/koor-tech/data-control-center/gen/go/api/services/cluster/v1/clusterv1connect"
	"github.com/koor-tech/data-control-center/pkg/ancientt"
	"github.com/koor-tech/data-control-center/pkg/config"
	"github.com/koor-tech/data-control-center/pkg/grpc/auth"
	k8s "github.com/koor-tech/data-control-center/pkg/k8s"
	k8scache "github.com/koor-tech/data-control-center/pkg/k8s/cache"
	"go.uber.org/fx"
	"go.uber.org/zap"
)

// Server is used to implement stats services.
type Server struct {
	clusterv1connect.ClusterServiceHandler

	readOnly  bool
	logger    *zap.Logger
	auth      *auth.GRPCAuth
	k         *k8s.K8s
	kc        *k8scache.Cache
	Namespace string

	ancientt *ancientt.Runner
}

type Params struct {
	fx.In

	Logger   *zap.Logger
	GrpcAuth *auth.GRPCAuth
	K8S      *k8s.K8s
	K8SCache *k8scache.Cache
	Cfg      *config.Config

	Ancientt *ancientt.Runner
}

func New(p Params) (*Server, error) {
	return &Server{
		readOnly:  p.Cfg.ReadOnly,
		logger:    p.Logger,
		auth:      p.GrpcAuth,
		k:         p.K8S,
		kc:        p.K8SCache,
		Namespace: p.Cfg.Namespace,
		ancientt:  p.Ancientt,
	}, nil
}

func (s *Server) RegisterService(g *gin.RouterGroup) {
	path, handler := clusterv1connect.NewClusterServiceHandler(s, connect.WithInterceptors(
		s.auth.NewAuthInterceptor(),
	))
	g.Any(path+"/*path", gin.WrapH(handler))
}

func (s *Server) GetKoorCluster(_ context.Context, _ *connect.Request[clusterpb.GetKoorClusterRequest]) (*connect.Response[clusterpb.GetKoorClusterResponse], error) {
	kc, _ := s.kc.GetKoorCluster(s.Namespace)

	res := connect.NewResponse(&clusterpb.GetKoorClusterResponse{
		KoorCluster: kc,
	})
	return res, nil
}

func (s *Server) GetResources(ctx context.Context, _ *connect.Request[clusterpb.GetResourcesRequest]) (*connect.Response[clusterpb.GetResourcesResponse], error) {
	var resources []*cephv1.Resource

	cephResources, err := s.k.GetYamlCephResources(ctx, s.Namespace)
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, fmt.Errorf("error caused by %w", err))
	}

	for _, resource := range cephResources {
		resources = append(resources, &cephv1.Resource{
			Name:      resource.Name,
			Namespace: resource.Namespace,
			Content:   resource.Content,
			Kind:      resource.Kind,
			Object:    resource.Object,
		})
	}

	res := connect.NewResponse(&clusterpb.GetResourcesResponse{
		Resources: &cephv1.Resources{
			Resources: resources,
		},
	})
	return res, nil
}

func (s *Server) SaveResources(ctx context.Context, req *connect.Request[clusterpb.SaveResourcesRequest]) (*connect.Response[clusterpb.SaveResourcesResponse], error) {

	requestResource := req.Msg.Resource
	resource := cephv1.Resource{
		Name:      requestResource.Name,
		Namespace: requestResource.Namespace,
		Content:   requestResource.Content,
		Kind:      requestResource.Kind,
		Object:    requestResource.Object,
	}

	if len(resource.Name) == 0 || len(resource.Content) == 0 {
		return nil, connect.NewError(connect.CodeInternal, fmt.Errorf("error caused by empty file"))
	}

	err := s.k.SaveResource(ctx, &resource)
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, fmt.Errorf("error caused by %w", err))
	}

	res := connect.NewResponse(&clusterpb.SaveResourcesResponse{
		Resource: &resource,
	})
	return res, nil
}

File: src/store/cluster.ts
import { ConnectError } from '@connectrpc/connect';
import { defineStore, type StoreDefinition } from 'pinia';
import { KoorCluster } from '~~/gen/ts/api/resources/koor/v1/koor_pb';
import {
    GetResourcesResponse,
    GetTroubleshootReportResponse,
    SaveResourcesRequest,
    SaveResourcesResponse,
} from '~~/gen/ts/api/services/cluster/v1/cluster_pb';

export interface ClusterState {}

export const useClusterStore = defineStore('cluster', {
    state: () => ({}) as ClusterState,
    persist: false,
    actions: {
        async getKoorCluster(): Promise<KoorCluster> {
            const { $grpc } = useNuxtApp();

            try {
                const resp = await $grpc.getClusterClient().getKoorCluster({});

                return resp.koorCluster!;
            } catch (e) {
                if (e instanceof ConnectError) $grpc.handleError(e as ConnectError);
                throw e;
            }
        },
        async getTroubleshootReport(): Promise<GetTroubleshootReportResponse> {
            const { $grpc } = useNuxtApp();

            try {
                return await $grpc.getClusterClient().getTroubleshootReport({});
            } catch (e) {
                if (e instanceof ConnectError) $grpc.handleError(e as ConnectError);
                throw e;
            }
        },
        async getCustomResources(): Promise<GetResourcesResponse> {
            const { $grpc } = useNuxtApp();
            try {
                return await $grpc.getClusterClient().getResources({});
            } catch (e) {
                if (e instanceof ConnectError) $grpc.handleError(e as ConnectError);
                throw e;
            }
        },
        async updateResource(
            name: string | undefined,
            namespace: string | undefined,
            kind: string | undefined,
            object: string | undefined,
            content: string,
        ): Promise<SaveResourcesResponse> {
            const { $grpc } = useNuxtApp();
            try {
                const updateResource = {
                    resource: { name, namespace, content, kind, object },
                } as SaveResourcesRequest;
                return await $grpc.getClusterClient().saveResources(updateResource);
            } catch (e) {
                if (e instanceof ConnectError) $grpc.handleError(e as ConnectError);
                throw e;
            }
        },
    },
});

if (import.meta.hot) {
    import.meta.hot.accept(acceptHMRUpdate(useClusterStore as unknown as StoreDefinition, import.meta.hot));
}
